### PR TITLE
Updated Course/Product saving logic

### PR DESCRIPTION
**Pull request by clintonb**
This logic now resides in the model, instead of the view. Moving the information here maintains a separation of concerns—the view shouldn't care how the model is saved—and clears up issues with the related product models and associated nested views.

XCOM-520

@AlasdairSwan @rlucioni @jimabramson @Nickersoft

### CODE REVIEW

File: ecommerce/static/js/models/course_model.js [hunk 1]
```diff
@@ -67,25 +77,182 @@ define([
...
+            /**
+             * Removes the parent products from the Product collection.
+             *
+             * This product is never exposed to the user, and should be ignored for all data operations.
+             */
+            removeParentProducts: function () {
+                var products = this.get('products'),
+                    parents = products.filter(function (product) {
+                        return product.get('structure') === 'parent';
+                    });
```
Comment (AlasdairSwan) [inline ecommerce/static/js/models/course_model.js]: Underscore's _.where function could work here (http://underscorejs.org/#where)

```
parents = _.where( products, {structure: 'parent');
```

File: ecommerce/static/js/models/course_model.js [hunk 2]
```diff
@@ -67,25 +77,182 @@ define([
...
+                    verification_deadline: null
+                };
+
+                // Submit only the relevant products
+                data.products = _.map(this.getCleanProducts(), function (product) {
+                    return product.toJSON();
+                }, this);
+
+                if (this.isIdVerified()) {
+                    var verificationDeadline = this.get('verification_deadline');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/models/course_model.js]: var should be declared at the top of the function

File: ecommerce/static/js/views/course_form_view.js [hunk 3]
```diff
@@ -394,12 +365,11 @@ define([
...
-                    activeSeatTypes,
-                    message,
-                    view,
-                    verificationDeadline,
-                    products = [];
+                var $buttons,
+                    $submitButton,
+                    btnDefaultText,
+                    self = this,
+                    btnSavingContent = '<i class="fa fa-spinner fa-spin"></i> ' + gettext('Saving...');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: a11y: please add aria-hidden="true" to the <i> element

File: ecommerce/static/js/views/course_form_view.js [hunk 4]
```diff
@@ -410,48 +380,45 @@ define([
...
-
-                data = {
-                    id: this.getFieldValue('id'),
-                    name: this.getFieldValue('name'),
-                    products: products,
-                    verification_deadline: null
-                };
+                // Disable all buttons by setting the attribute (for <button>) and class (for <a>)
+                $buttons.attr('disabled', 'disabled');
+                $buttons.addClass('disabled');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: jQuery chaining means you don't need to call $buttons twice, i.e.

```
$buttons.attr('disabled', 'disabled')
        .addClass('disabled');
```

File: ecommerce/static/js/views/course_form_view.js [hunk 5]
```diff
@@ -410,48 +380,45 @@ define([
...
-                    data: JSON.stringify(data),
-                    dataType: 'json',
-                    headers: {'X-CSRFToken': $.cookie('ecommerce_csrftoken')},
-                    method: 'POST',
-                    url: '/api/v2/publication/',
-                    success: function (data, textStatus, jqXHR) {
-                        this.goTo(data.id);
+                        // Re-enable the buttons
+                        $buttons.removeAttr('disabled');
+                        $buttons.removeClass('disabled');
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: See previous comment re. jQuery chaining

File: ecommerce/static/js/views/course_form_view.js [hunk 6]
```diff
@@ -410,48 +380,45 @@ define([
...
+                            message = response.responseJSON.error;
+                        }
+
                         // NOTE: jqXHR.responseJSON maps field names to an array of error messages.
-                        console.debug(jqXHR.responseJSON);
-                        message = jqXHR.responseJSON.error || gettext('An error occurred while saving the data.');
-                        this.clearAlerts();
-                        this.renderAlert('danger', message);
-                        this.$el.animate({scrollTop: 0}, 'slow');
+                        console.debug(response);
```
Comment (AlasdairSwan) [inline ecommerce/static/js/views/course_form_view.js]: Should this be committed to the repo? Or was it just for local development?

**Comment by AlasdairSwan** [comment]
Once tests pass and my minor comments have been addressed :+1:

File: ecommerce/static/js/views/course_form_view.js (hunk 6 above)
Comment (clintonb) [inline ecommerce/static/js/views/course_form_view.js]: Removed.

File: ecommerce/static/js/views/course_form_view.js (hunk 5 above)
Comment (clintonb) [inline ecommerce/static/js/views/course_form_view.js]: Updated.

File: ecommerce/static/js/views/course_form_view.js (hunk 4 above)
Comment (clintonb) [inline ecommerce/static/js/views/course_form_view.js]: Updated.

File: ecommerce/static/js/views/course_form_view.js (hunk 3 above)
Comment (clintonb) [inline ecommerce/static/js/views/course_form_view.js]: Updated.

File: ecommerce/static/js/models/course_model.js (hunk 2 above)
Comment (clintonb) [inline ecommerce/static/js/models/course_model.js]: Updated.

File: ecommerce/static/js/models/course_model.js (hunk 1 above)
Comment (clintonb) [inline ecommerce/static/js/models/course_model.js]: Updated.

**Pull request merged**